### PR TITLE
feat: migrate existing profiler metrics to generic metrics [MD-300] [MD-301]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

This PR migrates existing profiling metrics (system metrics only) in `trial_profiler_metrics` to generic metrics `metrics` and changes existing APIs related to the profiler to shim old APIs to fit the new schema.

Example old `trial_metrics_profiler` schema:
```
values: [0, 1, 2, 3]
batches: [0, 0, 1, 1]
ts: ["2023-06-23 09:26:50.443", "2023-06-23 09:26:50.443", "2023-06-23 09:26:50.443", "2023-06-23 09:26:50.443"]
labels: {"name": "free_memory", "agentId": "i-03b6ff93c18edf0ee", "trialId": 1, "metricType": "PROFILER_METRIC_TYPE_SYSTEM"}
```

Example new `metrics` schema:
```
trial: 1
end_time: "2023-06-23 09:26:50.443"
metrics: {
  "i-03b6ff93c18edf0ee": {
    "free_memory": 0
  }
}
```
The data in `trial_profiler_metrics` and table isn't dropped at this time, in case we need to rollback. Dropping the table will be done after this feature lands.

### Migration
Running the migration query took 1 minute on my local machine on a DB that had 55k rows in `trial_profiler_metrics`. This was the result of about 20 experiments, each ran with 2 agents and 8 GPUs per agent. 

It's hard to estimate the end user cost this will be, since it would require knowing how often people profile their experiments on how many agents/GPUs, and for how long. But this is something that will be called out when this is released.

## Test Plan

You should have access to the database you're testing on to make sure the data migration ran successfully. This can be roughly assessed with an equal count of the unique trial IDs in the previous table and new table:
```
select count(distinct labels->>'trialId') from trial_profiler_metrics;
select count(distinct trial_id) from metrics where partition_type='PROFILING';
```
After data migration is successful, there should be old metrics in the new `metrics` table partition. Find any trial that has profiling metrics from the old table and go to the web UI's "profiler" tab. The "Throughput" and "Timing metrics" should be empty (this will be deprecated by web UI in a separate PR) but the "system metrics" should render metrics and dropdowns for metrics, agents, and GPUs:

Screenshot 2024-03-06 at 3 54 22 PM

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/static/migrations/20240306144927_trial-profiler-metrics-migration.tx.up.sql [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    max(pm.ts) as end_time,
+    (
+        case pm.mgroup
+            when 'gpu' then jsonb_build_object(
+                    pm.agent_id,
+                    jsonb_object_agg(
+                        -- GPU UUID should never be null here, but SQL doesn't know that.
+                            coalesce(pm.gpu_uuid, ''), pm.metrics
+                        )
+                )
```
Comment (ioga) [inline master/static/migrations/20240306144927_trial-profiler-metrics-migration.tx.up.sql:21]: indentation here is a bit weird. I'd expect

```suggestion
            when 'gpu' then jsonb_build_object(
                pm.agent_id,
                jsonb_object_agg(
                    -- GPU UUID should never be null here, but SQL doesn't know that.
                    coalesce(pm.gpu_uuid, ''), pm.metrics
                )
            )
```


I remember Hamid added an SQL formatter: https://github.com/determined-ai/determined/pull/7538/files
maybe run it and see how it comes out?

File: master/static/migrations/20240306144927_trial-profiler-metrics-migration.tx.up.sql [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            else jsonb_object_agg(pm.agent_id, pm.metrics)
+            end
+        ) as metrics,
+    coalesce(t.run_id, 1) as trial_run_id,
+    'PROFILING' as partition_type,
+    pm.mgroup as metric_group
+from
+    (
+        -- Query to downsample existing metrics to one per metric name <> group <> agent <> trial <> gpu per second
```
Comment (ioga) [inline master/static/migrations/20240306144927_trial-profiler-metrics-migration.tx.up.sql:30]: cool

File: master/static/migrations/20240306144927_trial-profiler-metrics-migration.tx.up.sql [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                    when tpm.labels->>'name' like 'cpu\_%' then 'cpu'
+                    when tpm.labels->>'name'='free_memory' then 'memory'
+                    end
+                ) as mgroup,
+            (tpm.labels->>'trialId')::int as trial_id,
+            tpm.labels->>'agentId' as agent_id,
+            tpm.labels->>'gpuUuid' as gpu_uuid
+        from trial_profiler_metrics tpm
+                 -- Only get every 10 reported time, value pairs (currently reports every 0.1s)
+                 cross join lateral unnest(tpm.values, tpm.ts) with ordinality as vals(v, t, n)
```
Comment (ioga) [inline master/static/migrations/20240306144927_trial-profiler-metrics-migration.tx.up.sql:57]: cool

**Review by ioga — APPROVED** [review]

File: master/static/migrations/20240306144927_trial-profiler-metrics-migration.tx.up.sql (hunk 1 above)
Comment (azhou-determined) [reply master/static/migrations/20240306144927_trial-profiler-metrics-migration.tx.up.sql]: i ran that and it formatted just about everything _except_ this SQL file... in any case, i just formatted this by hand.

**Review by hamidzr — COMMENTED** [review]
lgtm but I think this now falls into @determined-ai/cluster-mgmt perhaps? If so they'd probably want to take a look as well. or maybe it's RM? 🤷🏼

**Comment by stoksc** [comment]
I assert that model dev owns these end to end. IDK what other think.

**Comment by stoksc** [comment]
Maybe slightly shared with exp tracking since they both use the metrics infra.

**Review by jgongd — APPROVED** [review]

**Pull request merged**